Okta.AspNet/DiscoveryDocumentCachingSigningKeyProvider.cs

```diff
@@ -0,0 +1,73 @@
+﻿// <copyright file="DiscoveryDocumentCachingSigningKeyProvider.cs" company="Okta, Inc">
+// Copyright (c) 2018-present Okta, Inc. All rights reserved.
+// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Okta.AspNet
+{
+    internal class DiscoveryDocumentCachingSigningKeyProvider
+    {
+        private readonly TimeSpan _refreshInterval = new TimeSpan(1, 0, 0, 0);
+        private readonly IDiscoveryDocumentSigningKeyProvider _discoveryDocumentSigningKeyProvider;
+        private DateTimeOffset _syncAfter = new DateTimeOffset(new DateTime(2001, 1, 1));
+        private IEnumerable<SecurityKey> _keys;
+
+        public DiscoveryDocumentCachingSigningKeyProvider(IDiscoveryDocumentSigningKeyProvider provider)
+        {
+            _discoveryDocumentSigningKeyProvider = provider ?? throw new ArgumentNullException(nameof(provider), "The provider cannot be null.");
+            RefreshMetadata();
+        }
+
+        /// <summary>
+        /// Gets all known security keys.
+        /// </summary>
+        /// <value>
+        /// All known security keys.
+        /// </value>
+        public IEnumerable<SecurityKey> SigningKeys
+        {
+            get
+            {
+                RefreshMetadata();
+                return _keys;
+            }
+        }
+
+        private void RefreshMetadata()
+        {
+            if (_syncAfter >= DateTimeOffset.UtcNow)
+            {
+                return;
+            }
+
+            // Queue a refresh, but discourage other threads from doing so.
+            _syncAfter = DateTimeOffset.UtcNow + _refreshInterval;
```

Comment:
I can see the intent here is to prevent multiple threads from executing this block simultaneously. I feel like this section should be protected by a lock to ensure only one thread executes this code at a time.  It should help to avoid potential anomalous behavior.  Let me know your thoughts here.

Reply:
Good point. I had the same thought at the beginning, but it seems that could also cause pool exhaustion and they removed locks in this PR: https://github.com/aspnet/AspNetKatana/pull/46. 
The current approach is the one that they are using for Active Directory.

Reply:
Good to know 👍